### ISSUE TITLE
useQuery in urql-vue re-executes query when other query get's executed

### ISSUE THREAD

**Issue by mlesin** [urql-graphql/urql#3150]
### Describe the bug

I have a reactive query that gets a list of items using useQuery from urql-vue, without using any conditions used in variables.
When I execute another query using useClientHandle, the previous query is executed again.
In my case, this behavior introduces an infinite loop, but I could not reproduce the loop in the example.
However, the second execution is reproducible:

1. Open the link provided in the "Reproduction" section
2. Note that there is a list of successfully fetched "todos", and during this "set: todos: count(5)" is printed in the "Log" section.
3. Press the "click me" button.
4. You will notice that along with the response in the Log section, another line of "set: todos: count(5)" appears.

### Reproduction

https://codesandbox.io/p/sandbox/gallant-dream-luddjc

### Urql version

@urql/core: 4.0.4
@urql/exchange-graphcache: 6.0.1
@urql/vue: 1.1.0

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/urql-graphql/urql/discussions) should be used
- [X] Read the [docs](https://formidable.com/open-source/urql/docs/).
- [X] Follow our [Code of Conduct](https://github.com/urql-graphql/urql/blob/main/CODE_OF_CONDUCT.md)

**Comment by mlesin** [comment]
In my case I'm getting a list with a short amount of properties in each object, and then receiving a full object after user selection, which causing list update, and, as I depend on that list too while detecting a selection change (for the case when update was initiated by subscription, for example), this triggers selection watcher again. And somewhere in the middle (I can't identify exactly where for a moment) I'm getting a loop.
Comparing lists to identify was there a real change or not could be pretty expensive operation if list is large.
Is there any chance to not to fire a list update if there was no change in any properties introduced as a result?

**Comment by kitten** [comment]
AH! ok, I understand. Yes, we probably need a shortcircuit like this, which makes sure we don't trigger the ref updates if they;re not needed: https://github.com/urql-graphql/urql/blob/35657e771633be43b850f0a1f7fdbac94b075dd7/packages/react-urql/src/hooks/state.ts#L39

That said, I don't currently remember if Vue's reactivity model supports "lensing" because you may have to also then take care thatyou don't recompute any updates if fetching/stale states change

**Comment by mlesin** [comment]
That means i have to perform comparison of two lists with deep analysis of their properties, that was exactly the type of expensive operation better to avoid.
I meant maybe it is a better idea to detect this at the level of cache update, when you only have two objects to compare (one in a query result's list, and another is fresh fetched object), and if no change in properties detected (only ones which are in the list already), do not touch that list at all?

**Comment by kitten** [comment]
You won't have to. Graphcache should keep objects and arrays deeply reference equal if no data has changed

**Comment by mlesin** [comment]
But when I added deep comparison manually in watcher, infinite loop I'm facing is gone. That means Graphcache is triggering watcher, but nothing was really changed in data array being watched. Is there any way to avoid ref to be notified about the change if no data actually changed?

**Comment by kitten** [comment]
Just to double-check here, I believe my point still makes sense in that I'd advise you to not rely on the watchers being ultra-reliable, and sometimes to call you even if nothing has changed or if something unrelated has changed. Generally, like in other bindings/frameworks/libraries, effects are reactive, but that doesn't always mean that it's trivial to tell why an event has been triggered or not.

That said, I think this is actually related to #3165, and after that was merged and released this fix should be causing data from Graphcache to remain referentially stable more often. Before, we had a slight issue in that it was accidentally reusing data from the API.

**Comment by mlesin** [comment]
Thanks a lot for your support!

**Comment by mlesin** [comment]
I repeated test case using new versions of packages here:
https://codesandbox.io/p/sandbox/cranky-tess-7x48qv

@urql/core: 4.0.6
@urql/exchange-graphcache: 6.0.2

Behavior is the same as before :(

I've added additional checks in my production code, so the main problem with the infinite loop went away at the cost of deep checking two (potentially large) arrays for equality.

**Comment by kitten** [comment]
<img width="1092" alt="Screenshot 2023-04-20 at 20 12 13" src="https://user-images.githubusercontent.com/2041385/233464679-2108c83d-f717-424f-a669-71dfd2091402.png">

To be honest, the only thing that it can now be is the deep proxy conversion that's happening. As far as I know, that's pretty idiomatic Vue code and a lot of people would expect this to be able to deeply react to changes in the data, but my guess is now that that's the culprit

**Comment by kitten** [comment]
I believe this may be a combination of two issues. #3231 and #3228 may together fix this issue. I'll check back once these are merged and ready to test

**Comment by kitten** [comment]
I'm not sure if this is something I've overlooked previously, but I actually just noticed that the `requestPolicy` is defaulted to `network-only`. This is causing the added request, since the update that Graphcache sees will also cause the `network-only` operation to be repeated, so this is expected behaviour 😅 Removing the `requestPolicy` works as expected.

Just noting that setting `network-only` also disables all caching, as per the docs, so the cache is ineffective when it's set. Hope that helps ✌️ 

<img width="293" alt="Screenshot 2023-06-01 at 22 45 46" src="https://github.com/urql-graphql/urql/assets/2041385/b207a9bf-0807-407f-a6b0-13550110573e">
